presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java

```diff
@@ -74,7 +74,15 @@ public LocalDispatchQuery(
         this.queryExecutor = requireNonNull(queryExecutor, "queryExecutor is null");
         this.querySubmitter = requireNonNull(querySubmitter, "querySubmitter is null");
 
-        addExceptionCallback(queryExecutionFuture, stateMachine::transitionToFailed);
+        // Check in the exception callback if we are already in the terminal state
+        // so that throwable could be ignored (CancellationException will be thrown
+        // when we cancel queryExecutionFuture in addStateChangeListener).
+        addExceptionCallback(queryExecutionFuture, throwable -> {
+            if (!stateMachine.isDone()) {
+                stateMachine.transitionToFailed(throwable);
+            }
+        });
```

Comment:
This LGTM. Please move to a separate PR

Reply:
Sure

Reply:
https://github.com/prestosql/presto/pull/2607